source/transfer_list.rst

```diff
@@ -141,15 +141,39 @@ The TE header is defined in :numref:`tab_te_header`.
      - The size of this entry header in bytes. This field is set to 8 for the TE header layout described in this version of the table.
 
 
-Entry type ranges
+Entry type allocation
 -----------------
 
-The content of the data section is determined by the tag id. The tag id space contains two ranges:
-
- #. Standard range, and
- #. Non-standard range
-
-The *tag_id* ranges are described in :numref:`tab_tag_id_ranges`.
+Tag IDs must be allocated in this specification before use. A new tag ID can be
+allocated by submitting a pull request to this repository that adds a description
+of the respective TE data layout to this specification. Tag IDs do not have to be
+allocated in order. Submitters are encouraged to try to group tag IDs together in
+logical clusters (e.g. all tags related to a particular chipset or to a
+particular firmware project could use adjacent tag numbers), but there are no
+predefined ranges and no reservations of tag ranges for specific use. The
+initially legal range for tag allocations is {0, ..., 0xf_ffff} (more ranges
+will be opened up when necessary).
+
+Tags are expected to have a simple layout and each tag should only represent
+data for a single logical concept. Data for multiple distinct concepts should be
+split across different tags, even if they're always expected to appear together
+on the first platform adding the tag (to encourage reusability in different
+situations). A list of objects can be represented by adding multiple TEs with
+the same tag ID to the TL.
+
+The {0x10_0000, ... 0x10_ffff} range is reserved for non-standardized use. Anyone
+is free to use tags from that range for any custom TE layout without adding their
+definitions to this specification first. The use of this range is *strongly
+discouraged* for anything other than local experiments or code that will only
+ever be used in closed-source components owned by the entity controlling the
+entire final firmware image. In particular, a TE just containing
+platform-specific data or internal structures specific to a single firmware
+implementation is no reason not to allocate a standardized tag for it in this
+specification. Since standards often emerge organically, the goal is to create
+unique tag IDs for everything just in case it turns out to be useful in more
+applications than initially anticipated. Basically, whenever you're submitting
+code for a new TE layout to any public open-source project, that's probably a
+good indication that you should allocate a tag ID for it in this specification.
```

Comment:
Well, there is no "how it's being used today" since the spec is not being used at all yet. If we're defining the non-standard tag space here, isn't this also the right place to explain how it should (and shouldn't) be used? If we put it in a different document, I'm worried that people may not see it when they decide how to implement their tags.

I see a very real risk with these two tag spaces that everyone is just going to create non-standard tags because it's easier, which defeats the benefit of having a unified specification because none of the actual contents of the transfer list can be unified between anything. Worse, when multiple independent stakeholders create components for the same firmware, they may create non-standard tags that end up overlapping, which would then probably lead to a lot of tiny separate tag space splitting agreements between those stakeholders outside of this spec... and if you then replaced one of the components in that firmware stack with one from a different firmware stack that's part of a different out-of-spec tag splitting agreement it's all going to be a huge mess.

I think the only way to make sure that we really retain interoperability and reusability for this is if everyone allocates all their tags in the standardized space and adds them to the spec. So I think it's really important to encourage that as much and as highly visible as possible, because the non-standard space already has the advantage that it's less work to use.

Reply:
By "how it's used today", I meant the specific range allocations (of which there are currently only 2). The "how to create new tags" seems like project rather than spec info to me, but I don't feel that strongly about it. What's more important is agreeing the policy / expected usage and I think it's better to discuss that on the list, which you've already started.

Reply:
You seem to be anticipating/encouraging a proliferation of TE types, which is fine but I wonder if we should then put these in a separate file? Then again I guess we can cross that problem when we get to it.

Reply:
The spec is likely to grow, but each tag needs to be properly documented in the spec, since that allowed people to understand what is in each TE.

I don't like the idea of a huge migration of objects from TF-A and coreboot (RAM stage onwards) to TEs. They should use device tree instead. But that is a separate discussion with each project.

Reply:
Yeah, if this actually sees widespread adoption it will definitely come to the point where it becomes unwieldy to define every tag in one file, so we'll want to decide on a better way to organize things eventually. I think the question of whether TE layout definitions should be C code instead kinda plays into that discussion as well... e.g. if we decided to do that we could leave the "spec" .rst document itself as just defining the header layouts and just put a link to the individual tag definitions (like an "appendix") that are all gathered in a directory with one header file per 256-wide tag range or something like that.

If we want to stick with keeping all in .rst I guess we can still have a main document that then refers to a directory of others, although I'm not sure how that would work with the PDF compilation thing then.

Reply:
OK, sounds like we should leave it as rST for now, but look at how we can add C code to the repo down the track. Just a separate licence, perhaps?

Reply:
Yes, if we add C code we should use a separate license (e.g. BSD-3-Clause). As long as this is clearly separate from the doc, it should be OK (though it would result in a bit of legal admin our side).